### DISCUSSION TITLE
I got problems while installing on Kaggle

### DISCUSSION THREAD

**Discussion by ghost** [BindsNET/bindsnet#664] [General]
I got problems while installing on Kaggle. Package dependencies are broken.

**Comment by Hananel-Hazan** [comment]
Can you be more specific? What kind of error did you get? what dependencies failed?

**Reply by ghost** [reply]
pip install git+https://github.com/BindsNET/bindsnet.git
!pip uninstall -y protobuf
!pip install protobuf==3.20.1

If you do this, then everything works. In principle, it is not neccesery to touch anything. It would be great if nothing changed in
in the near future (a week or two).

**Reply by Hananel-Hazan** [reply]
Thank you for bringing this issue up.
However, I'm still not sure what error you are getting.
The `protobuf` that is required by BinsNET is `4.25.1` which is the newest one.

Can you past here the error you encountered that require you to downgrade it to `3.20.1`

**Reply by ghost** [reply]
Attempting uninstall: torchaudio
    Found existing installation: torchaudio 2.0.1
    Uninstalling torchaudio-2.0.1:
      Successfully uninstalled torchaudio-2.0.1
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf 23.8.0 requires cupy-cuda11x>=12.0.0, which is not installed.
cuml 23.8.0 requires cupy-cuda11x>=12.0.0, which is not installed.
dask-cudf 23.8.0 requires cupy-cuda11x>=12.0.0, which is not installed.
apache-beam 2.46.0 requires dill<0.3.2,>=0.3.1.1, but you have dill 0.3.7 which is incompatible.
apache-beam 2.46.0 requires protobuf<4,>3.12.2, but you have protobuf 4.25.1 which is incompatible.
apache-beam 2.46.0 requires pyarrow<10.0.0,>=3.0.0, but you have pyarrow 11.0.0 which is incompatible.
cudf 23.8.0 requires pandas<1.6.0dev0,>=1.3, but you have pandas 2.0.3 which is incompatible.
cuml 23.8.0 requires dask==2023.7.1, but you have dask 2023.12.0 which is incompatible.
cuml 23.8.0 requires distributed==2023.7.1, but you have distributed 2023.12.0 which is incompatible.
dask-cudf 23.8.0 requires dask==2023.7.1, but you have dask 2023.12.0 which is incompatible.
dask-cudf 23.8.0 requires distributed==2023.7.1, but you have distributed 2023.12.0 which is incompatible.
dask-cudf 23.8.0 requires pandas<1.6.0dev0,>=1.3, but you have pandas 2.0.3 which is incompatible.
google-cloud-aiplatform 0.6.0a1 requires google-api-core[grpc]<2.0.0dev,>=1.22.2, but you have google-api-core 2.11.1 which is incompatible.
google-cloud-automl 1.0.1 requires google-api-core[grpc]<2.0.0dev,>=1.14.0, but you have google-api-core 2.11.1 which is incompatible.
google-cloud-bigquery 2.34.4 requires protobuf<4.0.0dev,>=3.12.0, but you have protobuf 4.25.1 which is incompatible.
google-cloud-bigtable 1.7.3 requires protobuf<4.0.0dev, but you have protobuf 4.25.1 which is incompatible.
google-cloud-pubsub 2.18.3 requires grpcio<2.0dev,>=1.51.3, but you have grpcio 1.51.1 which is incompatible.
google-cloud-vision 2.8.0 requires protobuf<4.0.0dev,>=3.19.0, but you have protobuf 4.25.1 which is incompatible.
kaggle-environments 1.14.3 requires gymnasium==0.29.0, but you have gymnasium 0.29.1 which is incompatible.
kaggle-environments 1.14.3 requires shimmy>=1.2.1, but you have shimmy 0.2.1 which is incompatible.
kfp 2.0.1 requires google-cloud-storage<3,>=2.2.1, but you have google-cloud-storage 1.44.0 which is incompatible.
kfp 2.0.1 requires protobuf<4,>=3.13.0, but you have protobuf 4.25.1 which is incompatible.
kfp-pipeline-spec 0.2.2 requires protobuf<4,>=3.13.0, but you have protobuf 4.25.1 which is incompatible.
tensorflow-metadata 0.14.0 requires protobuf<4,>=3.7, but you have protobuf 4.25.1 which is incompatible.
tensorflow-transform 0.14.0 requires protobuf<4,>=3.7, but you have protobuf 4.25.1 which is incompatible.
tensorflowjs 4.14.0 requires packaging~=23.1, but you have packaging 21.3 which is incompatible.
Successfully installed AutoROM.accept-rom-license-0.6.1 Cython-0.29.36 ale-py-0.8.1 autorom-0.4.2 bindsnet-0.3.2 cmake-3.27.9 eagerpy-0.30.0 foolbox-3.3.3 gymnasium-0.29.1 lit-17.0.6 nvidia-cublas-cu11-11.10.3.66 nvidia-cuda-cupti-cu11-11.7.101 nvidia-cuda-nvrtc-cu11-11.7.99 nvidia-cuda-runtime-cu11-11.7.99 nvidia-cudnn-cu11-8.5.0.96 nvidia-cufft-cu11-10.9.0.58 nvidia-curand-cu11-10.2.10.91 nvidia-cusolver-cu11-11.4.0.1 nvidia-cusparse-cu11-11.7.4.91 nvidia-nccl-cu11-2.14.3 nvidia-nvtx-cu11-11.7.91 protobuf-4.21.12 scikit-build-0.17.6 shimmy-0.2.1 tensorboardX-2.6.1 torch-2.0.1 torchaudio-2.0.2 torchvision-0.15.2 triton-2.0.0
Note: you may need to restart the kernel to use updated packages.

**Reply by ghost** [reply]
Here I dumped the stack trace.

**Reply by Hananel-Hazan** [reply]
Thanks, from the log above it seem that the issue you facing come from libraries that are not related to `BindsNET` :
- `protobuf <4` require for google-cloud-* services. 
-  `apache-beam` require older version of `dill` and older version of `pyarrow`
- there many other libraries that are not part of BindsNET

Those libraries prevent you to use newer dependencies that require for `BindsNET`. I don't see an easy way to fix this. I think that if you are using an older libraries, you need to find a workaround that fit to your specific environment.

For my curiosity:
What environment are you using? Conda? Venv?
What python version are you using?

**Reply by ghost** [reply]
I used what Kaggle provides by default when selecting the last environment. The Python version seemed to be lower than necessary. Honestly, I didn't delve into such details.

**Reply by Hananel-Hazan** [reply]
You may be succeeding with installing an older version of BindsNET that use older dependencies that can satisfy the google-cloud-* and the apache-beam services.  
I think the dependencies that google and apache complaining about come from newer PyTorch or other packages that BindsNET use.

**Reply by ghost** [reply]
ok, thank you

**Reply by ghost** [reply]
Can you answer, please? https://github.com/BindsNET/bindsnet/discussions/665
